sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,277 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities for ``FileSystem`` implementations."""
+
+import abc
+import io
+import os
+
+__all__ = ['Downloader', 'Uploader', 'DownloaderStream', 'UploaderStream',
+           'PipeStream']
+
+
+class Downloader(object):
+  """Download interface for a single file.
+
+  Implementations should support random access reads.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def size(self):
+    """Size of file to download."""
+
+  @abc.abstractmethod
+  def get_range(self, start, end):
+    """Retrieve a given byte range from this download, inclusive.
+
+    Range must be in this form:
+      0 <= start <= end: Fetch the bytes from start to end.
+
+    Args:
+      start: (int) Initial byte offset.
+      end: (int) Final byte offset, inclusive.
+
+    Returns:
+      (string) A buffer containing the requested data.
+    """
+
+
+class Uploader(object):
+  """Upload interface for a single file."""
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def last_error(self):
+    """Last error encountered for this instance."""
+
+  @abc.abstractmethod
+  def put(self, data):
+    """Write data to file sequentially.
+
+    Args:
+      data: (memoryview) Data to write.
+    """
+
+  @abc.abstractmethod
+  def finish(self):
+    """Signal to upload any remaining data and close the file.
+
+    File should be fully written upon return from this method.
+
+    Raises:
+      Any error encountered during the upload.
+    """
+
+
+class DownloaderStream(io.RawIOBase):
+  """Provides a stream interface for Downloader objects."""
+
+  def __init__(self, downloader, mode='r'):
+    """Initializes the stream.
+
+    Args:
+      downloader: (Downloader) Filesystem dependent implementation.
+      mode: (string) Python mode attribute for this stream.
+    """
+    self._downloader = downloader
+    self.mode = mode
+    self._position = 0
+
+  def readinto(self, b):
+    """Read up to len(b) bytes into b.
+
+    Returns number of bytes read (0 for EOF).
+
+    Args:
+      b: (bytearray/memoryview) Buffer to read into.
+    """
+    self._checkClosed()
+    if self._position >= self._downloader.size:
+      return 0
+
+    start = self._position
+    end = min(self._position + len(b) - 1, self._downloader.size - 1)
```

Comment:
<!--new_thread; commit:79815e6df1d49c271a6c44e37fb2a3c46be0aed5; resolved:0-->
See comment above about possibly making this range end non-inclusive.